crates/core/src/db/queries/collections.rs

```diff
@@ -0,0 +1,69 @@
+//! Query utilities for collections.
+
+use anyhow::Context;
+use diesel::{
+    pg::Pg,
+    sql_types::{Array, Integer, Nullable, Text},
+    types::ToSql,
+    RunQueryDsl,
+};
+
+use crate::{
+    db::{custom_types::OrderDirection, models::Nft, Connection},
+    error::Result,
+};
+
+/// Query collections ordered by volume
+///
+/// # Errors
+/// returns an error when the underlying queries throw an error
+pub fn by_volume(
+    conn: &Connection,
+    addresses: impl ToSql<Nullable<Array<Text>>, Pg>,
+    order_direction: OrderDirection,
+    limit: impl ToSql<Integer, Pg>,
+    offset: impl ToSql<Integer, Pg>,
+) -> Result<Vec<Nft>> {
+    diesel::sql_query(make_by_volume_query_string(order_direction))
+        .bind(addresses)
+        .bind(limit)
+        .bind(offset)
+        .load(conn)
+        .context("Failed to load collections by volume")
+}
+
+fn make_by_volume_query_string(order_direction: OrderDirection) -> String {
+    format!(
+        r"
+    SELECT
+        metadatas.address,
+        metadatas.name,
+        metadatas.seller_fee_basis_points,
+        metadatas.update_authority_address,
+        metadatas.mint_address,
+        metadatas.primary_sale_happened,
+        metadatas.uri,
+        metadatas.slot,
+        metadata_jsons.description,
+        metadata_jsons.image,
+        metadata_jsons.category,
+        metadata_jsons.model
+    FROM metadata_jsons
+    INNER JOIN metadatas ON (metadatas.address = metadata_jsons.metadata_address)
+    INNER JOIN (
+        SELECT metadata_collection_keys.collection_address AS collection, SUM(purchase_receipts.price) AS volume
+            FROM purchase_receipts
+            INNER JOIN metadatas on (purchase_receipts.metadata = metadatas.address)
+            INNER JOIN metadata_collection_keys on (metadatas.address = metadata_collection_keys.metadata_address)
+            WHERE ($1 IS NULL OR metadata_collection_keys.collection_address = ANY($1))
+            GROUP BY metadata_collection_keys.collection_address
+            ORDER BY volume {order_direction}
+            LIMIT $2
+            OFFSET $3
```

Comment:
Instead of re-running this query on each row query performance may be improved if assembly of the volume stats is done in its own query then references using WITH clause.

Reply:
Looks to improve the query by up to ~200 ms 🎉